src/irmin-pack/unix/dispatcher.ml

```diff
@@ -0,0 +1,255 @@
+open Import
+module Payload = Control_file.Latest_payload
+
+module type S = sig
+  module Fm : File_manager.S
+
+  type t
+
+  val v : root:string -> Fm.t -> (t, [> Fm.Errs.t ]) result
+
+  val read_exn : t -> off:int63 -> len:int -> bytes -> unit
+  (** [read_exn] either reads in the prefix or the suffix file, depending on
+      [off]. See [Io.read_exn] for the arguments. If it tries to read a gced
+      object, an exception is raised. *)
+
+  val read_at_most_exn : t -> off:int63 -> len:int -> bytes -> int
+  (** [read_at_most_exn] is similar to [read_exn] but if the end of file is
+      reached while reading [len] bytes, then only the available bytes are read.
+      No [`Read_out_of_bounds] error is raised. The number of bytes read are
+      returned. *)
+
+  val end_offset : t -> int63
+  (** [end_offset] is the end offsets of the pack entries, counting that the
+      prefix doesn't start at 0. It counts the entries not yet flushed from the
+      prefix. *)
+
+  val read_exn_if_not_gced : t -> off:int63 -> len:int -> bytes -> bool
+  (** Similar to [read_exn] but returns false if the object was gced, instead of
+      raising an expection. *)
+
+  val offset_of_suffix_off : t -> int63 -> int63
```

Comment:
optional idea: distinguish at the type level between "absolute" offsets and suffix offsets.

Reply:
tracked by https://github.com/mirage/irmin/issues/1954